cpu/esp_common/freertos/queue.c

```diff
@@ -110,13 +113,91 @@ void vQueueDelete( QueueHandle_t xQueue )
     free(xQueue);
 }
 
+BaseType_t IRAM_ATTR xQueueReset( QueueHandle_t xQueue )
+{
+    DEBUG("%s pid=%d queue=%p\n", __func__, thread_getpid(), xQueue);
+
+    assert(xQueue != NULL);
+
+    vTaskEnterCritical(0);
+
+    _queue_t* queue = (_queue_t*)xQueue;
+    queue->item_front = 0;
+    queue->item_tail = 0;
+    queue->item_level = 0;
+
+    /* return if there is no waiting sending thread */
+    if (queue->sending.next == NULL) {
+        DEBUG("%s pid=%d queue=%p return pdPASS\n", __func__,
+              thread_getpid(), xQueue);
+        vTaskExitCritical(0);
+        return pdPASS;
+    }
+
+    /* otherwise unlock the waiting sending thread */
+    list_node_t *next = list_remove_head(&queue->sending);
+    thread_t *proc = container_of((clist_node_t*)next, thread_t, rq_entry);
+    sched_set_status(proc, STATUS_PENDING);
+
+    /* test whether context switch is required */
+    bool ctx_switch = proc->priority < sched_threads[thread_getpid()]->priority;
+
+    DEBUG("%s pid=%d queue=%p unlock waiting pid=%d switch=%d\n",
+          __func__, thread_getpid(), xQueue, proc->pid, ctx_switch);
+
+    if (ctx_switch) {
+        vTaskExitCritical(0);
+        /* sets only the sched_context_switch_request in ISRs */
+        sched_switch(proc->priority);
+    }
+    else {
+        vTaskExitCritical(0);
+    }
```

Comment:
Wouldn't this work just as well?

```suggestion
    if (ctx_switch) {
        /* sets only the sched_context_switch_request in ISRs */
        sched_switch(proc->priority);
    }

    vTaskExitCritical(0);
```

Reply:
Critical sections for FreeRTOS are realized by disabling the interrupts. If the function is called from thread context and the context switch is executed with interrupts disabled, the interrupts are kept disabled. Therefore, in my opinion, they must be enabled before switching the context.